src/bloqade/builder/parse/trait.py

```diff
@@ -1,60 +1,170 @@
+
     def parse_circuit(self: "Builder") -> "AnalogCircuit":
-        """Parse the analog circuit from the program."""
+        """
+        Parse the analog circuit from the program.
+
+        Returns:
+            AnalogCircuit: The parsed analog circuit.
+
+        Raises:
+            ValueError: If the circuit cannot be parsed.
+        """
         from bloqade.builder.parse.builder import Parser
 
         return Parser().parse_circuit(self)
 
-
 class ParseRoutine:
+    """
+    A class providing functionality to parse the program and return a Routine object.
+
+    Example:
+        >>> class MyBuilder(ParseRoutine):
+        ...     pass
+        >>> builder = MyBuilder()
+        >>> routine = builder.parse()
+    """
+
     def parse(self: "Builder") -> "Routine":
-        """Parse the program to return a Routine object."""
+        """
+        Parse the program to return a Routine object.
+
+        Returns:
+            Routine: The parsed routine object.
+
+        Raises:
+            ValueError: If the routine cannot be parsed.
+        """
         from bloqade.builder.parse.builder import Parser
 
         return Parser().parse(self)
 
-
 class Parse(ParseRegister, ParseSequence, ParseCircuit, ParseRoutine):
+    """
+    A composite class inheriting from ParseRegister, ParseSequence, ParseCircuit, and ParseRoutine.
+    Provides a unified interface for parsing different components of the program.
+
+    Example:
+        >>> class MyBuilder(Parse):
+        ...     pass
+        >>> builder = MyBuilder()
+        >>> n_atoms = builder.n_atoms
+    """
+
     @property
-    def n_atoms(self: "Builder"):
-        """Return the number of atoms in the program."""
+    def n_atoms(self: "Builder") -> int:
+        """
+        Return the number of atoms in the program.
+
+        Returns:
+            int: The number of atoms in the parsed register.
 
+        Raises:
+            ValueError: If the register type is unsupported.
+
+        Note:
+            If the register is of type ParallelRegister, the number of atoms is extracted from its internal register.
+        """
         register = self.parse_register()
 
         if isinstance(register, ir.location.ParallelRegister):
             return register._register.n_atoms
         else:
             return register.n_atoms
 
-
 class Show:
+    """
+    A class providing functionality to display the builder with given arguments and batch ID.
+
+    Example:
+        >>> class MyBuilder(Show):
+        ...     pass
+        >>> builder = MyBuilder()
+        >>> builder.show()
+    """
+
     def show(self, *args, batch_id: int = 0):
+        """
+        Display the builder with the given arguments and batch ID.
+
+        Args:
+            *args: Additional arguments for display.
+            batch_id (int, optional): The batch ID to be displayed. Defaults to 0.
+
+        Note:
+            This method uses the `display_builder` function to render the builder's state.
```

Comment:
Since the `show` method will appear in the hint path, it would be nice to have the examples appear here instead of in the class docstring. Generally, we only need a little docstring for the builder class, which serves as a mixin. On the other hand, I think the example should demonstrate you can visualize the entire program at different stages.